### ISSUE TITLE
Slow to draw initial window, time spent in FcFontMatch

### ISSUE THREAD

**Issue by jonhoo** [jwilm/alacritty#1603]
Which operating system does the issue occur on? Arch Linux
If on linux, are you using X11 or Wayland? X11

Here's a brief screen capture of the issue: https://jon.thesquareplanet.com/share/alacritty-slow-start.mp4

Running it through [`callgrind`](http://valgrind.org/docs/manual/cl-manual.html), I get a profile that points pretty squarely at `FcFontMatch` being the culprit. Sadly, bumping the fontconfig and freetype dependencies doesn't fix the issue (#1602).

[![callgrind capture](https://jon.thesquareplanet.com/share/alacritty-slow-start-profile.png)](https://jon.thesquareplanet.com/share/alacritty-slow-start-profile.png)

Any suggestions for what to try next? Has anyone else seen a similar issue?

**Comment by chrisduerr** [comment]
Which GPU(+driver)/WM are you using? I'm on AMD+Mesa with i3 and I haven't seen issues where the startup time would be more than a fraction of a second. Such a big amount of time spent inside `FcFontMatch` seems like font fallback might be running into some problems.

Is this startup the same with a different font? Also has this been a recent regression or has this always been an issue?

**Comment by jwilm** [comment]
What is a typical startup time?

**Comment by jwilm** [comment]
Oops, I missed the gif.

**Comment by jonhoo** [comment]
This is a brand new Lenovo X1 Carbon (gen 6) with an Intel i7-8650U CPU and the associated integrated graphics, using the i915 driver. I don't think the specs should be the reason for any slowness. WM is xmonad. Running `time alacritty -e echo` on my laptop gives ~0.33s elapsed time. On my desktop (also xmonad, but discrete graphics), it's ~0.20s with the same font (Noto Sans Mono). Changing the font (to Liberation Mono) does not make a difference. Anything else I could try to help narrow it down?

Sadly, given that it's a brand new laptop, I can't really speak to whether it's a regression. I did try re-building `master` on my desktop now though, and launch time for `alacritty -e echo` remained the same (~0.20s), which suggests it's probably not a regression.

**Comment by jonhoo** [comment]
Okay, so let me backtrack a little -- launch time is improved a little by not running through `cargo run`. There's still a noticeable lag though, and the profiling still suggests that that is due to `FcFontMatch`. Now that I've seen it on my laptop, I'm also noticing it more on my desktop. It could be that there isn't anything out of the ordinary happening here, but simply that 0.3s is long enough to notice...

**Comment by jonhoo** [comment]
One thing that would help would be if the window immediately started receiving keyboard input, even before it has drawn anything (and before it does font stuff). That would reduce the pain, as if I start typing a command while the window is being drawn, those inputs wouldn't be missed.

**Comment by jonhoo** [comment]
Digging into the profile some more, it looks like `FcFontMatch` is called 14 times during startup; 8 times from `FreeTypeRasterizer::get_glyph` and 6 from `FreeTypeRasterizer::load_font`. Shouldn't nearly all of those hit the cache? Especially given that I'm using the _same_ font for normal, bold, and italic?

**Comment by chrisduerr** [comment]
Now this is interesting. Your benchmarks seem to suggest that the performance is the same it is on my machine and  your system definitely is more than powerful enough, however the gif looks far slower than my machine.

Just as a comparison, here's a gif of my machine: 
https://gfycat.com/DisgustingHoarseChameleon

Note that Alacritty seems to start and close significantly faster. And typing directly during startup, no key is lost. Using a Dvorak keyboard, so I was typing `aoeu` and the `a` was received before it started and `oeu` was sent to the new process, so nothing was lost.

Startup performance can definitely be improved. But there seems to be some discrepancy between the benchmark results and the actual perceived performance.

**Comment by jonhoo** [comment]
I feel like I'm becoming paranoid. I brought my `alacritty.yml` in-line with the upstream one, but basically keeping my settings ([commit here](https://github.com/jonhoo/configs/commit/f8d8de3f51c5481de155644297bbb88411207df9)), and that seems like it may have speeded things up a little further. Maybe it was because of disabling scrollback or something?

Yeah, I don't know how to explain it at this point. Here's a screencap from my laptop now without going through `cargo` and with the updated config: https://jon.thesquareplanet.com/share/alacritty-slow-ish.mp4. Seems a little better, and more in line with what you get.

I wonder if this could be related to the previous setup being on a rotated external screen... I'm at home now, but can check next time I'm in the office.

As an aside, I think this would *feel* a lot faster if alacritty drew the background color immediately...

**Comment by chrisduerr** [comment]
Scrollback actually shouldn't have a big impact on performance anymore. Alacritty's maximum amount of lines initialized at startup is the visible region plus 1000 lines. When your WM isn't resizing the window, it should be just the visible region.

The GIF you shared seems fine. Rendering the background color instead of an empty window would be an option though, however that might be something that needs to be tracked in winit/glutin.

You could try disabling the transparency built into Alacritty, that might draw the window as opaque instead of transparent at startup. That's just a guess though, iirc winit still starts the window as transparent initially.
[This is the code in question.](https://github.com/jwilm/alacritty/blob/master/src/window.rs#L370)

**Comment by jonhoo** [comment]
Looks like it still draws transparently initially unfortunately even with that line removed.

I _think_ this issue then reduces to 0.3s is long enough to notice, but that there isn't anything particularly wrong with my setup. My desktop is 0.2s, which is fast enough that I hadn't noticed before. I'll check the next time I can if there's a significant slowdown when using a rotated external monitor, which is where I initially noticed the issue.

Looking at the profiles though, I suspect we could cut down launch time for alacritty _significantly_ by optimizing the calls to fontconfig. In particular, I think it should only be necessary to call `FcFontMatch` a single time with my configuration, as opposed to the 14 times it is currently called during startup. It suggests to me that alacritty doesn't currently cache those lookups, and that if it did, that would be enough to make startup way faster.

**Comment by chrisduerr** [comment]
Okay, I'd then suggest changing the purpose of this issue from solving a bug, to improving startup performance by reducing the `FcFontMatch` calls. The title seems fine to me, just going to add some tags.

If there's any actual buggy behavior, it should be tracked in a different issue.

**Comment by jonhoo** [comment]
Yeah, I'll open a new issue if I discover that the external screen makes a difference.

Digging through, I get these primary call chains (probably incomplete due to inlining):

 - (`Display::new` + `Display::handle_resize` + `Display::draw`) -> `FreeTypeRasterizer::get_glyph` -> `FcFontMatch`
 - (`Display::new` + `Display::handle_resize`) -> `GlyphCache::compute_font_keys` -> `FreeTypeRasterizer::load_font` -> `FcFontMatch`

If I disable inlining on all `GlyphCache` methods, I get that the calls are coming from `GlyphCache::new`, `GlyphCache::update_font_size` (through `Display::new` and `Display::handle_resize` respectively). It looks like there are just many repeated (and unnecessary?) calls to `FcFontMatch` indirectly through the `GlyphCache` methods, when in reality those calls could probably all be reduced to a single one. My current profiles suggest that that could possibly be near a 10x speedup for startup.

**Comment by chrisduerr** [comment]
Thanks for looking into this. After the big speed improvement with dynamic allocation for large scrollback sizes, further improving Alacritty's boot time would really be nice.

If you want to take a stab at resolving this issue and have any questions, please let me know.

**Comment by jonhoo** [comment]
It looks like the extra time is spent primarily in `glutin::platform::platform::x11::Context::new`. Specifically in `glutin::api::glx::ffi::glx::Glx::QueryExtensionsString`.

[![profile with external monitor](https://jon.thesquareplanet.com/share/alacritty-slower-on-external.svg)](https://jon.thesquareplanet.com/share/alacritty-slower-on-external.svg)

**Comment by chrisduerr** [comment]
I'd like to keep discussions on this issue in #782, however this looks like a different problem which is not due to an issue with Alacritty, but a problem with glutin.

I'd recommend to test if this is reproducible with the glutin examples and report it upstream if it is.

**Comment by jonhoo** [comment]
Yeah, that's why I posted it here instead of in #782 :) I'll investigate with `glutin`.

**Comment by jonhoo** [comment]
Filed tomaka/glutin#1070
